keps/sig-network/2009-dns-autopath/README.md

```diff
@@ -0,0 +1,228 @@
+ 
+The new searchpath is of the form: 
+`search.$NS.$SUFFIX.ap.k8s.io`, where $NS is the namespace of the pod and $SUFFIX is the cluster suffix. ap.k8s.io is the delimiter to identify if a query needs search expansion.
+
+This searchpath can be set by using `dnsPolicy:None` to start with, as described [here](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pod-dns-config).
+Once this feature graduates to GA, the default searchpath for `ClusterFirst` and `ClusterFirstWithHostNet` will include this new searchpath instead of the 3 different k8s specific searchpaths.
+
+With this new searchpath, when a clientPod in the "default" namespace looks up a service - "mytestsvc”:
+
+1) The query will be sent out as `mytestsvc.search.default.cluster.local.ap.k8s.io`. 
+
+2) clusterDNS server(coreDNS by default) receives this and strips off the delimiter - "ap.k8s.io" and identifies "search" as start of the custom searchpath. The namespace and cluster suffix can be obtained from rest of the string. The DNS server can now construct the full service name for this query as - `mytestsvc.default.svc.cluster.local`. 
+
+This approach minimizes the number of DNS queries at client side to atmost 2(A, AAAA). The searchpath expansion logic moves to the server side.
+
+However, the above approach requires the clusterDNS server to know about the k8s DNS schema as well as the syntax of this custom searchpath. It does know about k8s DNS schema anyway, but knowledge of the custom searchpath is an additional requirement.
+
+### New EDNS0 Option
+
+As an alternative, we introduce an EDNS0 option which will include a list of
+searchpaths to be applied to the base query. 
+
+A new EDNS0 option - SearchPaths is introduced, with an option code from the [experimental range - 65001 to 65534](https://www.iana.org/assignments/dns-parameters/dns-parameters.xhtml#dns-parameters-11).
+The value of this option will be a comma-separated string consisting of all the searchpaths which are to be appended to the main query and looked up. This option can be useful outside of Kubernetes as well.
+
+
+Instead of having to modify all client pod images to insert an EDNS0 option in their requests, a new CoreDNS plugin "gensearchpaths" will be introduced. This will generate a list of search paths as an EDNS0 option to a given query.
+
+This plugin can be used by [Nodelocal DNSCache](https://github.com/kubernetes/enhancements/blob/master/keps/sig-network/0030-nodelocal-dns-cache.md), which is a daemonset running a per-node DNS Cache. It can also be used by a sidecar container attached to a pod, to selectively use this feature. The sidecar container will run a stripped-down version of CoreDNS with the above plugin.
```

Comment:
I think even as a sidecar, it's not straight forward to know the namespace of the enclosing pod.  There is [this post](https://stackoverflow.com/questions/46046110/how-to-get-the-current-namespace-in-a-pod), that suggests looking in `/var/run/secrets/kubernetes.io/serviceaccount/namespace`, but this seem kludgy to me.
Perhaps when the sidecar is created, an environment variable could be set.

Reply:
The sidecar will basically be a CoreDNS instance running "gensearchpaths" plugin. It will use the output from that to populate the new EDNS0 option. So, this sidecar determines the namespace by parsing the searchpath in /etc/resolv.conf which will be set to "search.<namespace>.<cluster suffix>.ap.k8s.io".

Good catch about determining namespace from a sidecar, i thought this was exposed in an env var already. In this case though, I think we can get awat with the limitation.